lib/streamaggr/streamaggr.go

```diff
@@ -708,83 +726,101 @@ func (a *aggregator) runFlusher(pushFunc PushFunc, alignFlushToInterval, skipInc
+		}
+
+		a.dedupFlush(dedupInterval, dedupTime.UnixMilli(), dedupIdx, flushIdx)
 
-		for tickerWait(t) {
-			if ignoreFirstIntervals > 0 {
-				a.flush(nil, interval, true)
+		if ct.After(flushDeadline) {
+			// It is time to flush the aggregated state
+			if alignFlushToInterval && skipIncompleteFlush && !isSkippedFirstFlush {
+				a.flush(nil, interval, flushDeadline.UnixMilli(), flushIdx)
+				isSkippedFirstFlush = true
+			} else if ignoreFirstIntervals > 0 {
+				a.flush(nil, interval, flushDeadline.UnixMilli(), flushIdx)
 				ignoreFirstIntervals--
 			} else {
-				a.flush(pushFunc, interval, true)
+				a.flush(pf, interval, flushDeadline.UnixMilli(), flushIdx)
 			}
-
-			if alignFlushToInterval {
-				select {
-				case <-t.C:
-				default:
-				}
+			for ct.After(flushDeadline) {
+				flushDeadline = flushDeadline.Add(interval)
 			}
 		}
-	} else {
-		alignedSleep(dedupInterval)
-		t := time.NewTicker(dedupInterval)
-		defer t.Stop()
-
-		flushDeadline := time.Now().Add(interval)
-		isSkippedFirstFlush := false
-		for tickerWait(t) {
-			a.dedupFlush(dedupInterval)
-
-			ct := time.Now()
-			if ct.After(flushDeadline) {
-				// It is time to flush the aggregated state
-				if alignFlushToInterval && skipIncompleteFlush && !isSkippedFirstFlush {
-					a.flush(nil, interval, true)
-					ignoreFirstIntervals--
-					isSkippedFirstFlush = true
-				} else if ignoreFirstIntervals > 0 {
-					a.flush(nil, interval, true)
-					ignoreFirstIntervals--
-				} else {
-					a.flush(pushFunc, interval, true)
-				}
-				for ct.After(flushDeadline) {
-					flushDeadline = flushDeadline.Add(interval)
-				}
-			}
-
-			if alignFlushToInterval {
-				select {
-				case <-t.C:
-				default:
-				}
+		if alignFlushToInterval {
+			select {
+			case <-t.C:
+			default:
 			}
 		}
 	}
 
 	if !skipIncompleteFlush && ignoreFirstIntervals <= 0 {
-		a.dedupFlush(dedupInterval)
-		a.flush(pushFunc, interval, true)
+		dedupTime := time.Now().Truncate(tickInterval).Add(tickInterval)
+		if a.ignoreOldSamples {
+			dedupIdx, flushIdx = a.getFlushIndices(dedupInterval, interval, dedupTime, flushDeadline)
+		}
+		a.dedupFlush(dedupInterval, flushDeadline.UnixMilli(), dedupIdx, flushIdx)
+		a.flush(pushFunc, interval, flushDeadline.UnixMilli(), flushIdx)
+	}
+}
+
+func (a *aggregator) getFlushIndices(dedupInterval, interval time.Duration, dedupTime, flushTime time.Time) (int, int) {
+	flushTimestamp := flushTime.UnixMilli()
+	flushIntervals := int(flushTimestamp / int64(interval/time.Millisecond))
+	var dedupIndex, flushIndex int
+	if dedupInterval > 0 {
+		dedupTimestamp := dedupTime.UnixMilli()
+		dedupIntervals := int(dedupTimestamp / int64(dedupInterval/time.Millisecond))
+		intervalsRatio := int(interval / dedupInterval)
+		dedupIndex = dedupIntervals % aggrStateSize
+		flushIndex = flushIntervals % (aggrStateSize / intervalsRatio)
+	} else {
+		flushIndex = flushIntervals % aggrStateSize
+		dedupIndex = flushIndex
 	}
+	return dedupIndex, flushIndex
```

Comment:
Running this locally for a while I've noticed that `dedupIndex` will always be `0` due to using ration between dedup and aggregation intervals (I've used `10s` and `20s` respectively). 
An updated version just uses `numberOfIntervals % aggrStateSize` and it works consistently as far as I can see.

```suggestion
	flushTimestamp := flushTime.UnixMilli()
	flushIntervals := int(flushTimestamp / int64(interval/time.Millisecond))
	flushIndex := flushIntervals % aggrStateSize
	dedupIndex := flushIndex

	if dedupInterval > 0 {
		dedupTimestamp := dedupTime.UnixMilli()
		dedupIntervals := int(dedupTimestamp / int64(dedupInterval/time.Millisecond))
		dedupIndex = dedupIntervals % aggrStateSize
	}
	return dedupIndex, flushIndex
```